### PR TITLE
Add checkpoint before additive model training so no rerun coalesce

### PR DESCRIPTION
**Pull request by jq**
@deerzq @saurfang

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 1]
```diff
@@ -124,15 +126,21 @@ object AdditiveModelTrainer {
...
-
-    input(params.subsample)
+    val data = input(params.subsample)
       .coalesce(params.numBags, true)
-      .mapPartitionsWithIndex((index, partition) => sgdPartition(index, partition, modelBC, sgdParams))
+
+    if (!params.checkPointDir.isEmpty) {
+      sc.setCheckpointDir(params.checkPointDir)
+      data.checkpoint()
+    }
```
Comment (saurfang) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: `sc` can be retrieved directly via `data.context`

**Comment by saurfang** [comment]
I am not sure if this will necessarily help you because transformed examples are one order of magnitude larger so they are slower to write and read in the event of failure, on the other hand checkpointing does potentially avoid shuffle the step in the event of failure. It is hard to say. 
If we see SGD fails often especially due to OOM, we can look into mini-batch SGD instead where gradient is updated per epoch so it can be computed in parallel.

**Comment by jq** [comment]
@saurfang Thanks for review, updated!

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 2]
```diff
@@ -56,11 +61,10 @@ object AdditiveModelTrainer {
...
                                    epsilon: Double, // epsilon used in epsilon-insensitive loss for regression training
-                                   initModelPath: String,
-                                   linearFeatureFamilies: java.util.List[String],
-                                   priors: Array[String],
-                                   nDTreePipelineParams: NDTreePipelineParams,
-                                   classWeights: Map[Int, Float])
+                                   init: InitParams,
+                                   shuffle: Boolean,
+                                   classWeights: Map[Int, Float],
+                                   checkPointDir: String)
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: we are adding more and more parameters here. you may want to write some comments to describe what are these

**Comment by deerzq** [comment]
Please add detailed documentation, then LGTM

**Pull request merged**